### PR TITLE
[Nix] package snarkyjs, client_sdk, mina-signer

### LINKED ISSUE #10077: [Nix] Frontend packaging
We have a JS/wasm target in one of the branches that should also be packaged with Nix. Furthermore, we should use buildYarnPackage to build the frontend code.

Branch: https://github.com/MinaProtocol/mina/tree/fix/kimchi-proof-js
build:
```
dune b src/lib/snarky_js_bindings/snarky_js_node.bc.js
dune b src/lib/snarky_js_bindings/snarky_js_chrome.bc.js
```

### PR DESCRIPTION
**Pull request by yorickvP**
The Nix code was still missing builds for client_sdk, snarkyjs and mina-signer. This PR:
- [x] Enables the Rust wasm builds (also needed a Cargo.lock)
- [x] Fixes the build for client_sdk_bindings
- [x] Adds a build for mina-signer (was missing `jest` as a dependency)
- [x] Adds a build for snarkyjs

Things left to do:
- [x] Fix https://github.com/serokell/nix-npm-buildpackage/issues/33 (worked around it in this code, but this is very ugly and mixes multiple npm versions)
- [ ] Make `snarky-run` work with this package
- [ ] Move snarkyjs tests into snarkyjs repo? @mitschabaude ?
- [ ] Run snarkyjs tests

* Closes #10077

### CODE REVIEW

File: nix/overlay.nix [hunk 1]
```diff
@@ -157,4 +158,97 @@ in {
...
+  # Work around https://github.com/rust-lang/wg-cargo-std-aware/issues/23
+  kimchi-rust-std-deps = pkgs.rustPlatform.importCargoLock {
+    # IFD:
+    lockFileContents = import "${
+        (pkgs.runCommand "std-deps" { } ''
+          mkdir $out
+          cp ${pkgs.kimchi-rust-wasm}/lib/rustlib/src/rust/Cargo.lock $out/
+          echo builtins.readFile ./Cargo.lock > $out/std-deps.nix
+        '')
+      }/std-deps.nix";
```
[the remark this answers is not in the document]
Comment (yorickvP) [reply nix/overlay.nix:182]: Try it and see :)

**Comment by mitschabaude** [comment]
>  Move snarkyjs tests into snarkyjs repo? @mitschabaude ?

No, I'd prefer if they stay where they are
* the ones in /tests depend only on the Mina code and are run as unit tests in Mina CI, so make more sense in that repo
* the ones in /test_module are for integration / unit tests in Mina CI; also depend on a local version of mina-signer which shouldn't be in snarkyjs (snarkyjs is supposed to work as a self-contained repo)

File: src/lib/crypto/kimchi_bindings/stubs/dune [hunk 2]
```diff
@@ -19,6 +19,7 @@
  (deps
   Cargo.toml
   (source_tree src)
+  (source_tree binding_generation)
```
Comment (yorickvP) [inline src/lib/crypto/kimchi_bindings/stubs/dune:22]: I should check if this doesn't break incremental builds

File: nix/overlay.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (balsoft) [reply nix/overlay.nix:182]: Ok, sure, still:
```nix
    lockFileContents = builtins.readFile (pkgs.runCommand "Cargo.lock" { }
      "cp ${pkgs.kimchi-rust-wasm}/lib/rustlib/src/rust/Cargo.lock $out");
```

**Review by balsoft — APPROVED** [review]

**Review by mrmr1993 — COMMENTED** [review]
Not really a fan of this. Can you make a version that doesn't change what dune is doing?

File: flake.nix [hunk 3]
```diff
@@ -172,6 +175,43 @@
           '';
         };
 
+        # snarkyjs
+        packages.snarky_js = nix-npm-buildPackage.buildNpmPackage {
+          src = ./src/lib/snarky_js_bindings/snarkyjs;
+          preBuild = ''
+            BINDINGS_PATH=./dist/server/node_bindings
```
[the remark this answers is not in the document]
Comment (yorickvP) [reply flake.nix:182]: It's currently inside `./scripts/build-snarkyjs-node.sh`, but it should probably be somewhere else, possibly on the ocaml side.

**Comment by yorickvP** [comment]
@mrmr1993 I'd like to talk about this some more. I basically need a Cargo.lock file for the `src/lib/crypto` libs.
The current strategy has.. a problem, in that the cargo target path is sometimes `.` and sometimes `./_build/default/src/lib/crypto/target`, depending on if dune copies over `src/lib/crypto/Cargo.toml`.
So I'd need about 3 cargo lockfiles with the current codebase, and I'd need to keep them in sync. Another thing we can do with the lockfile is switch to a cargo git dep instead of a git submodule :).

Would you be open to discuss ways to improve this situation?

**Review by mrmr1993 — APPROVED** [review]

**Pull request merged**